adapters/src/controller/mod.rs

```diff
@@ -0,0 +1,698 @@
+    /// `config` and starts streaming data from the endpoint if the pipeline
+    /// is running.
+    ///
+    /// # Errors
+    ///
+    /// The method may fail for the following reasons:
+    ///
+    /// * The endpoint configuration is invalid, e.g., specifies an unknown
+    ///   transport or data format.
+    ///
+    /// * The endpoint fails to initialize, e.g., because the network address or
+    ///   filename specified in the transport config is unreachable.
+    pub fn connect_input(
+        &self,
+        endpoint_name: &str,
+        config: &InputEndpointConfig,
+    ) -> AnyResult<()> {
+        self.inner.connect_input(
+            endpoint_name,
+            config,
+            self.circuit_thread_unparker.clone(),
+            self.backpressure_thread_unparker.clone(),
+        )?;
+        self.unpark_backpressure();
+        Ok(())
+    }
+
+    /// Change the state of all input endpoints to running.
+    ///
+    /// Start streaming data through all connected input endpoints.
+    pub fn start(&self) {
+        self.inner.start();
+        self.unpark_backpressure();
+    }
+
+    /// Pause all input endpoints.
+    ///
+    /// Sends a pause command to all input endpoints.  Upon receiving the
+    /// command, the endpoints must stop pushing data to the pipeline.  This
+    /// method is asynchronous and may return before all endpoints have been
+    /// fully paused.
+    pub fn pause(&self) {
+        self.inner.pause();
+        self.unpark_backpressure();
+    }
+
+    /// Terminate the controller, stop all input endpoints and destroy the
+    /// circuit.
+    pub fn stop(self) -> AnyResult<()> {
+        self.inner.stop();
+        self.unpark_circuit();
+        self.unpark_backpressure();
+        self.circuit_thread_handle
+            .join()
+            .map_err(|_| AnyError::msg("circuit thread panicked"))??;
+        self.backpressure_thread_handle
+            .join()
+            .map_err(|_| AnyError::msg("backpressure thread panicked"))?;
+        Ok(())
+    }
+
+    /// Unpark the circuit thread.
+    fn unpark_circuit(&self) {
+        self.circuit_thread_unparker.unpark();
+    }
+
+    /// Unpark the backpressure thread.
+    fn unpark_backpressure(&self) {
+        self.backpressure_thread_unparker.unpark();
+    }
+
+    /// Circuit thread function.
+    fn circuit_thread(
+        mut circuit: DBSPHandle,
+        controller: Arc<ControllerInner>,
+        parker: Parker,
+        backpressure_thread_unparker: Unparker,
+    ) -> AnyResult<()> {
+        let mut start: Option<Instant> = None;
+
+        let config = controller.config.read().unwrap();
+        let max_buffering_delay = Duration::from_micros(config.global.max_buffering_delay_usecs);
+        let min_batch_size_records = config.global.min_batch_size_records;
+        drop(config);
+
+        loop {
+            match controller.state.load(Ordering::Acquire) {
```

Comment:
If the status flag update gets reordered with thread unparking, this may cause the thread to sleep forever although it has work to do.

Reply:
Atomics can't be reordered relative to (un)parking, they can't be reordered around effectful operations like syscalls

Reply:
I looked at the implementation of `unpark`.  It uses atomics internally with `SeqCst`, which I think means that you're right.  There are other situations where weak consistency is problematic, though.  In particular, the unparker is only invoked when the number of buffered records changes from `< threshold` to `>= threshold` (or the other way around, depending on the context). I am not able to reliably reason about how weak consistency will affect the correctness of this logic.

In general, the way I am using atomics here (namely per batch rather than something more fine grained, e.g., per record), they are extremely unlikely to have any performance impact, so I don't think we need to worry about the cost of consistency.